api/src/controllers/login.js

```diff
@@ -229,10 +251,64 @@ const getBranding = () => {
     });
 };
 
+const renderTokenLogin = (req, res) => {
+  return getBranding()
+    .then(branding => render('tokenLogin', req, branding, { tokenUrl: req.url }))
+    .then(body => res.send(body));
+};
+
+const createSessionRetry = (req, retry= 10) => {
+  return createSession(req).then(sessionRes => {
+    if (sessionRes.statusCode === 200) {
+      return sessionRes;
+    }
+
+    if (retry) {
+      return createSessionRetry(req, retry - 1);
+    }
+
+    throw { status: 449, message: 'Login failed after 10 retries' };
+  });
+};
+
+/**
+ * Generates a session cookie for a user identified by supplied token and hash request params.
+ * The user's password is reset in the process.
+ */
+const tokenLogin = (req, res) => {
+  if (!req.params || !req.params.token || !req.params.hash) {
+    return res.status(400).json({ error: 'Missing required params' });
+  }
+
```

Comment:
Check `settings` immediately - if token login is disabled globally then exit right away. This gives an easy kill switch for admins to disable token login for all. Also I think most projects will disable token login so it closes off this API quickly so we can't get DOSed as easily.